Modules/_testcapi/object.c

```diff
@@ -485,6 +485,41 @@ is_uniquely_referenced(PyObject *self, PyObject *op)
 }
 
 
+static PyObject *
+object_setattr_null_exc(PyObject *self, PyObject *args)
+{
+    PyObject *obj, *name;
+    if (!PyArg_ParseTuple(args, "OO", &obj, &name)) {
+        return NULL;
+    }
+
+    PyErr_SetString(PyExc_ValueError, "error");
+    if (PyObject_SetAttr(obj, name, NULL) < 0) {
+        return NULL;
+    }
+    assert(PyErr_Occurred());
+    return NULL;
+}
+
+
+static PyObject *
+object_setattrstring_null_exc(PyObject *self, PyObject *args)
+{
+    PyObject *obj;
+    const char *name;
+    if (!PyArg_ParseTuple(args, "Os", &obj, &name)) {
```

Comment:
Use `z#`. So you can test with undecodable name -- this will allow to test that the value is tested for NULL in PyObject_SetAttrString, not only in PyObject_SetAttr, and that it is tested before attempting to decode the name.